### PR TITLE
feat: status chan

### PR DESCRIPTION
**Pull request by LeoLiuYan**

### CODE REVIEW

**Comment by LeoLiuYan** [comment]
cc @cpuguy83

**Comment by LeoLiuYan** [comment]
This pull request add `NodeControllerReady` and `PodControllerReady` function which expose in cli, is that ok?
@cpuguy83

**Comment by cpuguy83** [comment]
Can we instead expose the underlying controllers directly?

e.g. `cmd.NodeController()` and `cmd.PodController()`, and the readiness can be checked from there.

**Comment by LeoLiuYan** [comment]
> Can we instead expose the underlying controllers directly?
> 
> e.g. `cmd.NodeController()` and `cmd.PodController()`, and the readiness can be checked from there.

That's may be not appropriately since `PodCtroller` in virtual-kubelet include a sync.Mutex field.  So, should expose a interface when create `PodController` or `NodeController`?

**Comment by cpuguy83** [comment]
The mutex is not exposed, it is private to the package.

**Comment by LeoLiuYan** [comment]
> The mutex is not exposed, it is private to the package.

image
As the lint says,   "copylocks: assignment copies lock value" error, because of `sync.Mutext`.

**Comment by cpuguy83** [comment]
That's because you have it copying a value instead of passing a pointer.

File: internal/commands/root/root.go [hunk 1]
```diff
@@ -201,6 +204,9 @@ func runRootCommandWithProviderAndClient(ctx context.Context, pInit provider.Ini
 	if err != nil {
 		return errors.Wrap(err, "error setting up pod controller")
 	}
+	if podCtrl != nil {
+		*podCtrl = *pc
```
Comment (cpuguy83) [inline internal/commands/root/root.go:208]: Here, and similar with `nodeCtrl`, it is assigning the value of `pc` to the value of `podCtrl`.

This is where the linting error comes from.

Comment (LeoLiuYan) [reply internal/commands/root/root.go]: Already fix this issue.

**Comment by LeoLiuYan** [comment]
> That's because you have it copying a value instead of passing a pointer.

Already fix this issue.

**Comment by cpuguy83** [comment]
The root command is what creates the node and pod controller. There shouldn't be any need to pass that down.

**Comment by LeoLiuYan** [comment]
> The root command is what creates the node and pod controller. There shouldn't be any need to pass that down.

But the root command not return controllers, so pass a pointer when we create the command, and then assign the value to the pointer.

**Comment by cpuguy83** [comment]
What I'm saying is, let's have the command provide accessors for those controllers.

**Comment by LeoLiuYan** [comment]
Simplify the implements, remove point of controllers, just provide accessors. 
@cpuguy83

File: internal/commands/root/root.go [hunk 2]
```diff
@@ -42,6 +42,13 @@ import (
 	"k8s.io/client-go/tools/record"
 )
 
+var (
+	// podController is the pod controller which created by NewPodController.
+	podController *node.PodController
```
Comment (cpuguy83) [inline internal/commands/root/root.go:47]: We cannot use global vars like this.
These need to be plumbed through from whatever is returned from `NewCommand`.

If that means we need to return extra things or wrap `*cobra.Command` to add accessors, that's fine.

Comment (LeoLiuYan) [reply internal/commands/root/root.go]: Ok... review again.
@cpuguy83

**Comment by feiskyer** [comment]
Close since there's no activity for more than 5 months. Feel free to open a new PR if you think the changes are still required.

**Pull request closed without merging**